src/expr/skolem_manager.cpp

```diff
@@ -50,43 +51,91 @@ Node SkolemManager::mkSkolem(Node v,
+    d_gens[q] = pg;
+  }
   return getOrMakeSkolem(w, prefix, comment, flags);
 }
 
-Node SkolemManager::mkSkolemExists(Node v,
-                                   Node q,
-                                   const std::string& prefix,
-                                   const std::string& comment,
-                                   int flags,
-                                   ProofGenerator* pg)
+Node SkolemManager::mkSkolemize(Node q,
+                                std::vector<Node>& skolems,
+                                const std::string& prefix,
+                                const std::string& comment,
+                                int flags,
+                                ProofGenerator* pg)
+{
+  Trace("sk-manager-debug") << "mkSkolemize..." << std::endl;
+  Assert(q.getKind() == EXISTS);
+  Node currQ = q;
+  for (const Node& av : q[0])
+  {
+    Assert(currQ.getKind() == EXISTS && av == currQ[0][0]);
+    Node sk = skolemize(currQ, currQ, prefix, comment, flags);
+    Trace("sk-manager-debug")
+        << "made skolem " << sk << " for " << av << std::endl;
+    skolems.push_back(sk);
+  }
+  if (pg != nullptr)
+  {
+    // notice this may overwrite an existing proof generator
+    d_gens[q] = pg;
+  }
+  return currQ;
+}
+
+Node SkolemManager::skolemize(Node q,
+                              Node& qskolem,
+                              const std::string& prefix,
+                              const std::string& comment,
+                              int flags)
 {
   Assert(q.getKind() == EXISTS);
-  bool foundVar = false;
+  Node v;
   std::vector<Node> ovars;
+  std::vector<Node> ovarsW;
+  Trace("sk-manager-debug") << "mkSkolemize..." << std::endl;
+  NodeManager* nm = NodeManager::currentNM();
   for (const Node& av : q[0])
   {
-    if (av == v)
+    if (v.isNull())
     {
-      foundVar = true;
+      v = av;
       continue;
     }
+    // must make fresh variable to avoid shadowing, which is unique per
+    // variable av to ensure that this method is deterministic
+    Node avp = getOrMakeBoundVariable(av, av);
+    ovarsW.push_back(avp);
     ovars.push_back(av);
   }
-  if (!foundVar)
-  {
-    Assert(false);
-    return Node::null();
-  }
+  Assert(!v.isNull());
   Node pred = q[1];
+  qskolem = q[1];
+  Trace("sk-manager-debug") << "make exists predicate" << std::endl;
   if (!ovars.empty())
   {
-    NodeManager* nm = NodeManager::currentNM();
-    Node bvl = nm->mkNode(BOUND_VAR_LIST, ovars);
+    Node bvl = nm->mkNode(BOUND_VAR_LIST, ovarsW);
     pred = nm->mkNode(EXISTS, bvl, pred);
+    // skolem form keeps the old variables
+    bvl = nm->mkNode(BOUND_VAR_LIST, ovars);
+    qskolem = nm->mkNode(EXISTS, bvl, pred);
   }
-  return mkSkolem(v, pred, prefix, comment, flags, pg);
+  Trace("sk-manager-debug") << "call sub mkSkolem" << std::endl;
+  // don't use a proof generator, since this may be an intermediate, partially
+  // skolemized formula.
```

Comment:
Clarify that the proof generator for the whole skolemized formula is used is set in `mkSkolemize`